### PR TITLE
i#6668: Replace drwrap retaddr sentinel on all architectures, not just x86

### PR DESCRIPTION
**Pull request by ivankyluk**
Expand drwrap's return address sentinel replacement from just x86 (as was added in PR #6395) to cover all architectures.

When -record_replace_retaddr is used in ARM (as well as x86) platforms, the return address in the stack is replaced by the sentinel.
For a tailcall, the current implementation uses the address in the stack, the sentinel, as the return address. The change is to check if the return address in the stack is the sentinel or not. If it is, replace it with the return address of the outer level.

The problem can be reproduced by

./build/bin64/drrun -t drcachesim -offline -record_replace_retaddr -record_function 'tailcall_with_retaddr|1&foo|1' -- ../build_suite/build_debug-internal-64/suite/tests/bin/common.getretaddr

on an ARM machine.

This change expends PR #6395  to cover ARM.

Issues: #6394, #6668
Fixes: #6668

### CODE REVIEW

File: ext/drwrap/drwrap.c [hunk 1]
```diff
@@ -2021,9 +2021,8 @@ drwrap_in_callee(void *arg1, reg_t xsp _IF_NOT_X86(reg_t lr))
     NOTIFY(2, "%s: level %d function " PFX "\n", __FUNCTION__, pt->wrap_level + 1, pc);
```
Comment (derekbruening) [inline ext/drwrap/drwrap.c:2021]: > Fixes: https://github.com/DynamoRIO/dynamorio/issues/6668

I'd add another line `Issues: #6394, #6668`

**Review by derekbruening — APPROVED** [review]

File: ext/drwrap/drwrap.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (ivankyluk) [reply ext/drwrap/drwrap.c:2021]: Done.

[the remark this answers is not in the document]
Comment (ivankyluk) [reply ext/drwrap/drwrap.c:2021]: Done.

[the remark this answers is not in the document]
Comment (ivankyluk) [reply ext/drwrap/drwrap.c:2021]: Done.

Comment (ivankyluk) [reply ext/drwrap/drwrap.c:2021]: Done.

[the remark this answers is not in the document]
Comment (ivankyluk) [reply ext/drwrap/drwrap.c:2021]: Done.

File: suite/tests/CMakeLists.txt [hunk 2]
```diff
@@ -4777,7 +4777,7 @@ if (BUILD_CLIENTS)
     endif (BUILD_PT_TRACER AND BUILD_PT_POST_PROCESSOR)
   endif (proc_supports_pt)
 
-  if (X86)
+  if ((ARM AND NOT ANDROID) OR X86)
```
[the remark this answers is not in the document]
Comment (ivankyluk) [reply suite/tests/CMakeLists.txt]: Here's the results:

aarch64-precommit (ubuntu-20-arm64-sve):
102/109 Test #310: code_api|tool.drcacheoff.getretaddr_record_replace_retaddr .......   Passed    0.20 sec

aarch64-precommit (ubuntu-20-arm64):
90/108 Test #306: code_api|tool.drcacheoff.getretaddr_record_replace_retaddr .......   Passed    0.38 sec

x86-32:
427/491 Test #415: code_api|tool.drcacheoff.getretaddr_record_replace_retaddr .......   Passed    2.56 sec

**Pull request merged**